File: app/src/test/java/com/jbrunton/mymovies/fixtures/RepositoryFixtures.java
package com.jbrunton.mymovies.fixtures;

import com.jbrunton.mymovies.api.repositories.HttpMoviesRepository;
import com.jbrunton.entities.Movie;

import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;

import static org.mockito.Mockito.when;

public class RepositoryFixtures {
    public static class FakeMoviesRepositoryDsl {
        protected final HttpMoviesRepository repository;

        private FakeMoviesRepositoryDsl(HttpMoviesRepository repository) {
            this.repository = repository;
        }
    }

    public static class FakeMoviesFindDsl extends FakeMoviesRepositoryDsl {
        private final String id;

        public FakeMoviesFindDsl(HttpMoviesRepository repository, String id) {
            super(repository);
            this.id = id;
        }

        public void toReturn(Movie movie) {
            toReturnDelayed(movie, 0);
        }

        public void toReturnDelayed(Movie movie, int delay) {
            when(repository.getMovie(id)).thenReturn(Observable.just(movie).delay(delay, TimeUnit.SECONDS));
        }

        public void toErrorWith(Throwable throwable) {
            toErrorWithDelayed(throwable, 0);
        }

        public void toErrorWithDelayed(Throwable throwable, int delay) {
            when(repository.getMovie(id)).thenReturn(Observable.<Movie>error(throwable).delay(delay, TimeUnit.SECONDS));
        }
    }

    public static FakeMoviesFindDsl stubFind(HttpMoviesRepository repository, String id) {
        return new FakeMoviesFindDsl(repository, id);
    }
}

File: app/src/test/java/com/jbrunton/mymovies/fixtures/RepositoryFixtures.kt
package com.jbrunton.mymovies.fixtures

import com.jbrunton.async.AsyncResult
import com.jbrunton.entities.models.Movie
import com.jbrunton.entities.repositories.MoviesRepository
import com.nhaarman.mockito_kotlin.whenever
import io.reactivex.Observable
import java.util.concurrent.TimeUnit

object RepositoryFixtures {
    open class FakeMoviesRepositoryDsl constructor(protected val repository: MoviesRepository)

    class FakeMoviesFindDsl(repository: MoviesRepository, private val id: String) : FakeMoviesRepositoryDsl(repository) {

        fun toReturn(movie: Movie) {
            toReturnDelayed(movie, 0)
        }

        fun toReturnDelayed(movie: Movie, delay: Int) {
            val result: AsyncResult<Movie> = AsyncResult.Success(movie)
            whenever(repository.getMovie(id)).thenReturn(Observable.just(result).delay(delay.toLong(), TimeUnit.SECONDS))
        }

        fun toErrorWith(throwable: Throwable) {
            toErrorWithDelayed(throwable, 0)
        }

        fun toErrorWithDelayed(throwable: Throwable, delay: Int) {
            val result: AsyncResult<Movie> = AsyncResult.Failure(throwable, null)
            whenever(repository.getMovie(id)).thenReturn(Observable.just(result).delay(delay.toLong(), TimeUnit.SECONDS))
        }
    }

    fun stubFind(repository: MoviesRepository, id: String): FakeMoviesFindDsl {
        return FakeMoviesFindDsl(repository, id)
    }
}
